### PR TITLE
#2284, #2310

### PR DESCRIPTION
**Pull request by MikhailArkhipov**
#2284 Remoteness should be displayed visibly at all times
#2310 Connection manager tries to pass remote URL as local

### CODE REVIEW

File: src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs [hunk 1]
```diff
@@ -214,8 +214,11 @@ internal class ConnectionManager : IConnectionManager {
 
         private bool IsValidLocalConnection(string name, string path) {
             try {
-                var info = new RInterpreterInfo(name, path);
-                return info.VerifyInstallation();
+                Uri uri;
+                if (Uri.TryCreate(path, UriKind.Absolute, out uri) && uri.IsFile) {
```
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: So if this method returns `false` for remote connection, `LastActiveConnection` will be set to null and user will be switched from remote connection to the first available local. Is it a desired behavior?

File: src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs [hunk 2]
```diff
@@ -14,39 +15,67 @@
...
 using Microsoft.VisualStudio.Utilities;
+using static System.FormattableString;
 
 namespace Microsoft.VisualStudio.R.Package.Repl {
     [Export(typeof(IInteractiveWindowComponentContainerFactory))]
     internal class VsRInteractiveWindowComponentContainerFactory : IInteractiveWindowComponentContainerFactory {
         private readonly Lazy<IVsInteractiveWindowFactory> _vsInteractiveWindowFactoryLazy;
         private readonly IContentTypeRegistryService _contentTypeRegistryService;
+        private IRInteractiveWorkflow _interactiveWorkflow;
+        private IVsWindowFrame _frame;
```
Comment (int19h) [inline src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs]: Is the factory supposed to be mapped 1:1 to the entities that it is creating? It looks like you can only call `Create` once on this if you want it to work correctly.

File: src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs [hunk 3]
```diff
@@ -14,39 +15,67 @@
...
             _contentTypeRegistryService = contentTypeRegistryService;
+
         }
 
-        public IInteractiveWindowVisualComponent Create(int instanceId, IInteractiveEvaluator evaluator) {
+        public IInteractiveWindowVisualComponent Create(int instanceId, IInteractiveEvaluator evaluator, IRInteractiveWorkflow interactiveWorkflow) {
             VsAppShell.Current.AssertIsOnMainThread();
 
+            _interactiveWorkflow = interactiveWorkflow;
+            _interactiveWorkflow.RSessions.BrokerChanged += OnBrokerChanged;
```
Comment (AlexanderSher) [inline src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs]: I think we can just save the reference to `interactiveWorkflow.RSessions`

File: src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs [hunk 4]
```diff
@@ -14,39 +15,67 @@
...
+
+            UpdateWindowTitle();
 
             var componentContainer = new VisualComponentToolWindowAdapter<IInteractiveWindowVisualComponent>(toolWindow);
             var component = new RInteractiveWindowVisualComponent(vsWindow.InteractiveWindow, componentContainer);
             componentContainer.Component = component;
             return component;
         }
+
+        private void OnBrokerChanged(object sender, EventArgs e) {
```
Comment (AlexanderSher) [inline src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs]: I'd move the handler to the `RInteractiveWindowVisualComponent` itself, and unsubscribe it when visual component is disposed.

File: src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs (hunk 1 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: It is called `IsValidLocalConnection` after all... It didn't work for remote connections anyways. Currently CM does not restore remote connection - which should probably be filed as a separate bug.

File: src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs (hunk 2 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs]: Probably not (at least not eventually with multiple REPLs). Let me see what I can do.

File: src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs [hunk 5]
```diff
@@ -14,39 +15,67 @@
...
+            UpdateWindowTitle();
+        }
+
+        private void UpdateWindowTitle() {
+            var broker = _interactiveWorkflow.RSessions.Broker;
+            string title;
+            if (broker != null) {
+                title = broker.IsRemote
+                    ? Invariant($"{Resources.ReplWindowName} - {broker.Name} ({broker.Uri.ToString().TrimTrailingSlash()})")
```
Comment (int19h) [inline src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs]: This makes it look like an MDI parent window, or VS top level window itself (the "document - app" form). I think we need some other format to clearly disambiguate. `@` is traditionally used to indicate a network context - as in "Package Manager @ Remote1" etc.

File: src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs (hunk 1 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: Or should we just remove `IsValidLocalConnection` check for the MRU connection?

File: src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs (hunk 5 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs]: It is not something unseen before... 

image

File: src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs (hunk 1 above)
Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: Let's file another bug then

File: src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs (hunk 2 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs]: OK, separated frame operations and moved to the component

File: src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs (hunk 4 above)
Comment (MikhailArkhipov) [inline src/Package/Impl/Repl/VsRInteractiveWindowComponentContainerFactory.cs]: Done

File: src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs (hunk 1 above)
Comment (MikhailArkhipov) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: I fixed it, it now restores connection. But there is no indication in the UI that something is going on for ~5-10 seconds especially when REPL is closed.

Comment (AlexanderSher) [inline src/R/Components/Impl/ConnectionManager/Implementation/ConnectionManager.cs]: I'm not sure how good it is to show progress bar right on top of the VS that has just started. We can have an indication on a status bar.

**Comment by AlexanderSher** [comment]
:+1:

**Comment by int19h** [comment]
:+1:

**Pull request merged**